File: java/src/main/java/io/binac/leetcode/NQueensII.java
package io.binac.leetcode;

/**
 * The n-queens puzzle is the problem of placing n queens on an n×n chessboard such that no two queens attack each other.
 * <p>
 * <p>https://leetcode.com/static/images/problemset/8-queens.png
 * <p>
 * <p>Given an integer n, return the number of distinct solutions to the n-queens puzzle.
 * <p>
 * <p>Example:
 * <blockquote><pre>
 *     Input: 4
 *     Output: 2
 *     Explanation: There are two distinct solutions to the 4-queens puzzle as shown below.
 *     [
 *      [".Q..",  // Solution 1
 *       "...Q",
 *       "Q...",
 *       "..Q."],
 *
 *      ["..Q.",  // Solution 2
 *       "Q...",
 *       "...Q",
 *       ".Q.."]
 *     ]
 * </blockquote></pre>
 */
public class NQueensII {
    public static class Solution1 {
        private void totalNQueens(int n, int i, int[] count,
                                  boolean[] column, boolean[] mainDiagonal, boolean[] antiDiagonal) {
            if (i == n) {
                ++count[0];
                return;
            }
            for (int j = 0, d, a; j < n; ++j) {
                d = i - j + n - 1;
                a = i + j;
                if (column[j] || mainDiagonal[d] || antiDiagonal[a]) continue;
                column[j] = true;
                mainDiagonal[d] = true;
                antiDiagonal[a] = true;
                totalNQueens(n, i + 1, count, column, mainDiagonal, antiDiagonal);
                column[j] = false;
                mainDiagonal[d] = false;
                antiDiagonal[a] = false;
            }
        }

        public int totalNQueens(int n) {
            int[] result = new int[1];
            totalNQueens(n, 0, result, new boolean[n], new boolean[2 * n - 1], new boolean[2 * n - 1]);
            return result[0];
        }
    }
}

File: python3/leetcodepy/n_queens_ii.py
"""
The n-queens puzzle is the problem of placing n queens on an n x n chessboard such that no two queens attack each other.

Given an integer n, return the number of distinct solutions to the n-queens puzzle.



Example 1:


Input: n = 4
Output: 2
Explanation: There are two distinct solutions to the 4-queens puzzle as shown.

Example 2:

Input: n = 1
Output: 1


Constraints:

1 <= n <= 9
"""
from typing import List


class Solution1:
    def totalNQueens(self, n: int) -> int:
        result = [0]
        col_used = [False] * n
        diagonal_used = [False] * (2 * n - 1)
        anti_diagonal_used = [False] * (2 * n - 1)
        self.__total_n_queens(n, 0, col_used, diagonal_used, anti_diagonal_used, result)
        return result[0]

    def __total_n_queens(self, n: int, i: int,
                         col_used: List[bool], diagonal_used: List[bool], anti_diagonal_used: List[bool],
                         result: List[int]):
        if i == n:
            result[0] += 1
            return
        for j in range(n):
            diagonal, anti_diagonal = n - 1 - i + j, i + j
            if col_used[j] or diagonal_used[diagonal] or anti_diagonal_used[anti_diagonal]:
                continue
            col_used[j] = diagonal_used[diagonal] = anti_diagonal_used[anti_diagonal] = True
            self.__total_n_queens(n, i + 1, col_used, diagonal_used, anti_diagonal_used, result)
            col_used[j] = diagonal_used[diagonal] = anti_diagonal_used[anti_diagonal] = False
